nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/field/ObjectLocalDateTimeFieldConverter.java

```diff
@@ -67,22 +70,59 @@ public LocalDateTime convertField(final Object field, final Optional<String> pat
                 try {
                     return LocalDateTime.parse(string, formatter);
                 } catch (final DateTimeParseException e) {
-                    throw new FieldConversionException(LocalDateTime.class, field, name, e);
+                    return tryParseAsNumber(string, name);
                 }
             } else {
-                try {
-                    final long number = Long.parseLong(string);
-                    final Instant instant = Instant.ofEpochMilli(number);
-                    return ofInstant(instant);
-                } catch (final NumberFormatException e) {
-                    throw new FieldConversionException(LocalDateTime.class, field, name, e);
-                }
+                return tryParseAsNumber(string, name);
             }
         }
 
         throw new FieldConversionException(LocalDateTime.class, field, name);
     }
 
+    private LocalDateTime tryParseAsNumber(final String value, final String fieldName) {
+        // If decimal, treat as a double and convert to seconds and nanoseconds.
+        if (value.contains(".")) {
+            final double number = Double.parseDouble(value);
+            return toLocalDateTime(number);
+        }
+
+        // attempt to parse as a long value
+        try {
+            final long number = Long.parseLong(value);
+            return toLocalDateTime(number);
+        } catch (final NumberFormatException e) {
+            throw new FieldConversionException(LocalDateTime.class, value, fieldName, e);
+        }
+    }
+
+    private LocalDateTime toLocalDateTime(final double secondsSinceEpoch) {
+        // Determine the number of micros past the second by subtracting the number of seconds from the decimal value and multiplying by 1 million.
+        final long micros = (long) (1_000_000 * (secondsSinceEpoch - (long) secondsSinceEpoch));
+        // Convert micros to nanos. Note that we perform this as a separate operation, rather than multiplying by 1_000,000,000 in order to avoid
+        // issues that occur with rounding at high precision.
+        final long nanos = micros * 1000L;
+
+        return toLocalDateTime((long) secondsSinceEpoch, nanos);
+    }
+
+    private LocalDateTime toLocalDateTime(final long epochSeconds, final long nanosPastSecond) {
+        final Instant instant = Instant.ofEpochSecond(epochSeconds).plusNanos(nanosPastSecond);
+        return ofInstant(instant);
+    }
+
+    private LocalDateTime toLocalDateTime(final long value) {
+        final Instant instant = Instant.ofEpochMilli(value);
+        final LocalDateTime localDateTime = ofInstant(instant);
+        if (localDateTime.getYear() > 10_000) {
```

Comment:
Instead of comparing the year value after the conversion, what about checking that the initial `long value` is greater than a certain number equivalent to 10K years in the future? That would avoid multiple calls to `ofInstant()` in the case of handling microseconds.